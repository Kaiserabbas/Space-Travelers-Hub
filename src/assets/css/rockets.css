.rocket {
  display: flex;
  gap: 1em;
  margin: 1em 1.5em;
}

.rocket img {
  width: 250px;
  height: 250px;
}

.rocket-body {
  display: flex;
  flex-direction: column;
  gap: 0.5em;
  font-weight: 600;
}

.rocket-body span {
  background: rgb(2, 146, 2);
  margin-right: 0.5em;
  padding: 0.125em 0.25em;
  border-radius: 3px;
  color: #fff;
  font-weight: 500;
  font-size: 0.8em;
}

.rocket-body button {
  width: 30%;
  padding: 0.5em;
  background: rgb(45, 107, 231);
  color: white;
}
