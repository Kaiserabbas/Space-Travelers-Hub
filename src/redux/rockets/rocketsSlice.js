import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const initialState = {
  rockets: [],
  status: 'idle',
  error: null,
};

const fetchRockets = createAsyncThunk('rockets/fetchRockets', async () => {
  const response = await fetch('https://api.spacexdata.com/v4/rockets');
  const data = await response.json();
  return data.map((rocket) => ({
    id: rocket.id,
    rocket_name: rocket.name,
    type: rocket.type,
    flickr_images: rocket.flickr_images,
    description: rocket.description,
  }));
});

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const { id } = action.payload;
      state.rockets = state.rockets.map((rocket) => (
        rocket.id === id ? { ...rocket, reserved: true } : rocket));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.rockets = action.payload;
      });
  },
});

export { fetchRockets };
export const { reserveRocket } = rocketsSlice.actions;
export default rocketsSlice.reducer;
