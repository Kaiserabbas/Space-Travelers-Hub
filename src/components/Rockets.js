import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets, reserveRocket, cancelReservation } from '../redux/rockets/rocketsSlice';
import '../assets/css/rockets.css';

function Rockets() {
  const status = useSelector((state) => state.rockets.status);
  const rockets = useSelector((state) => state.rockets.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchRockets());
    }
  }, [status, dispatch]);

  const handleReserveClick = (rocketId) => {
    dispatch(reserveRocket({ id: rocketId }));
  };

  const handleCancelReservationClick = (rocketId) => {
    dispatch(cancelReservation({ id: rocketId }));
  };

  return (
    <>
      {rockets.map((rocket) => (
        <div className="rocket" key={rocket.id}>
          <img src={rocket.flickr_images[0]} alt="Rocket" />
          <div className="rocket-body">
            <h2>{rocket.rocket_name}</h2>
            {rocket.reserved ? (
              <>
                <p>
                  <span>Reserved</span>
                  {rocket.description}
                </p>
                <button
                  type="button"
                  onClick={() => handleCancelReservationClick(rocket.id)}
                  style={{
                    border: '1.5px solid gray',
                    background: 'transparent',
                    color: 'gray',
                  }}
                >
                  Cancel Reservation
                </button>
              </>
            ) : (
              <>
                <p>
                  {rocket.description}
                </p>
                <button
                  type="button"
                  onClick={() => handleReserveClick(rocket.id)}
                >
                  Reserve Rocket
                </button>
              </>
            )}
          </div>
        </div>
      ))}
    </>

  );
}

export default Rockets;
