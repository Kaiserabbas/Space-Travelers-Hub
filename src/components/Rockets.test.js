import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Rockets from './Rockets';
import { reserveRocket } from '../redux/rockets/rocketsSlice';

const mockStore = configureMockStore([thunk]);

describe('Rockets Component', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      rockets: {
        rockets: [
          {
            id: '5e9d0d95eda69955f709d1eb',
            rocket_name: 'Falcon 1',
            reserved: false,
          },
          {
            id: '5e9d0d95eda69973a809d1ec',
            rocket_name: 'Falcon 9',
            reserved: true,
          },
        ],
        status: 'succeeded',
      },
    });
  });

  test('Renders rockets and Reserve Rocket button', async () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    const rocketNameElement = screen.getByText('Falcon 1');
    const reserveButtonElement = screen.getByText('Reserve Rocket');

    expect(rocketNameElement).toBeInTheDocument();
    expect(reserveButtonElement).toBeInTheDocument();
  });

  test('Clicking Reserve Rocket dispatches reserveRocket action', async () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    const reserveButtonElement = screen.getByText('Reserve Rocket');
    fireEvent.click(reserveButtonElement);

    expect(store.getActions()).toContainEqual(reserveRocket({ id: '5e9d0d95eda69955f709d1eb' }));
  });
});
