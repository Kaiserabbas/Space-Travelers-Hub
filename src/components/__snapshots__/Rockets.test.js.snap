// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Component Rockets component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rocket"
  >
    <img
      alt="Rocket"
      src=""
    />
    <div
      class="rocket-body"
    >
      <h2>
        Falcon 1
      </h2>
      <p />
      <button
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    class="rocket"
  >
    <img
      alt="Rocket"
      src=""
    />
    <div
      class="rocket-body"
    >
      <h2>
        Falcon 9
      </h2>
      <p>
        <span>
          Reserved
        </span>
      </p>
      <button
        style="border: 1.5px solid gray; background: transparent; color: gray;"
        type="button"
      >
        Cancel Reservation
      </button>
    </div>
  </div>
</DocumentFragment>
`;
